File: node/vedadb-orm/src/vector.ts
/**
 * VedaDB ORM VectorProxy
 *
 * Provides vector similarity search via VedaDB VECTOR commands.
 */

import { DriverResult } from './types';
import { SchemaDefinition } from './schema';
import { escapeIdentifier } from './utils';

export interface VectorSearchOptions {
  field?: string;
  topK?: number;
  threshold?: number;
  filter?: Record<string, any>;
}

export class VectorProxy<T = any> {
  private schema: SchemaDefinition<T>;
  private executor: (sql: string) => Promise<DriverResult>;
  private hydrator: (result: DriverResult) => T[];

  constructor(
    schema: SchemaDefinition<T>,
    executor: (sql: string) => Promise<DriverResult>,
    hydrator: (result: DriverResult) => T[]
  ) {
    this.schema = schema;
    this.executor = executor;
    this.hydrator = hydrator;
  }

  /**
   * Find nearest neighbors for a given vector.
   */
  async search(vector: number[], options: VectorSearchOptions = {}): Promise<(T & { _distance?: number })[]> {
    const table = escapeIdentifier(this.schema.tableName);
    const vectorStr = `[${vector.join(', ')}]`;
    const topK = options.topK || 10;

    // Determine vector field
    let field = options.field;
    if (!field) {
      for (const [name, def] of Object.entries(this.schema.fields)) {
        if (def.type.toString() === 'VECTOR') {
          field = name;
          break;
        }
      }
    }
    if (!field) field = 'embedding';

    let sql = `SELECT *, VECTOR_DISTANCE(${escapeIdentifier(field)}, '${vectorStr}') AS _distance FROM ${table}`;

    // Optional filter
    if (options.filter && Object.keys(options.filter).length > 0) {
      const conditions = Object.entries(options.filter)
        .map(([k, v]) => {
          if (v === null) return `${escapeIdentifier(k)} IS NULL`;
          if (typeof v === 'string') return `${escapeIdentifier(k)} = '${v.replace(/'/g, "''")}'`;
          return `${escapeIdentifier(k)} = ${v}`;
        })
        .join(' AND ');
      sql += ` WHERE ${conditions}`;
    }

    if (options.threshold !== undefined) {
      const whereOrAnd = sql.includes('WHERE') ? 'AND' : 'WHERE';
      sql += ` ${whereOrAnd} VECTOR_DISTANCE(${escapeIdentifier(field)}, '${vectorStr}') <= ${options.threshold}`;
    }

    sql += ` ORDER BY _distance ASC LIMIT ${topK};`;

    const result = await this.executor(sql);
    return this.hydrator(result) as (T & { _distance?: number })[];
  }

  /**
   * Upsert a vector for a given record id.
   */
  async upsert(id: any, vector: number[], field?: string): Promise<void> {
    const table = escapeIdentifier(this.schema.tableName);
    const pk = escapeIdentifier(this.schema.primaryKeyField);
    const vecField = escapeIdentifier(field || 'embedding');
    const vectorStr = `[${vector.join(', ')}]`;

    await this.executor(
      `UPDATE ${table} SET ${vecField} = '${vectorStr}' WHERE ${pk} = '${String(id).replace(/'/g, "''")}';`
    );
  }
}

File: python/vedadb_orm/vector.py
"""
VedaDB ORM — VectorProxy.

Wraps VedaDB's vector-similarity search and returns typed model instances
alongside their distance/similarity scores.
"""

from __future__ import annotations

import json
from typing import Any, Dict, List, Optional, Tuple, TYPE_CHECKING

from .exceptions import QueryError
from .utils import format_where_value

if TYPE_CHECKING:
    pass


class VectorProxy:
    """Model-scoped vector search builder."""

    def __init__(
        self,
        model_cls: type,
        embedding: List[float],
        top_k: int = 10,
        metric: Optional[str] = None,
    ) -> None:
        self._model_cls = model_cls
        self._embedding = embedding
        self._top_k = top_k
        self._metric = metric
        self._filters: Dict[str, Any] = {}

    def filter(self, **where: Any) -> "VectorProxy":
        """Add equality filters applied *before* the vector search."""
        self._filters.update(where)
        return self

    def exec(self) -> List[Tuple[Any, float]]:
        """Execute the vector search.

        Returns a list of ``(model_instance, score)`` tuples ordered by
        similarity.
        """
        orm = self._model_cls._get_orm()
        if orm is None:
            raise QueryError("Model is not bound to a VedaORM instance")

        table = self._model_cls._table_name
        schema = getattr(self._model_cls, "__schema__", None)

        vec_field = None
        metric = self._metric
        if schema and schema.vector.enabled:
            vec_field = schema.vector.field
            if metric is None:
                metric = schema.vector.metric

        if vec_field is None:
            raise QueryError("No vector field configured for this model")

        if metric is None:
            metric = "cosine"

        vec_json = json.dumps(self._embedding)

        sql = (
            f"SELECT *, VECTOR_DISTANCE({vec_field}, '{vec_json}', '{metric}') AS _score "
            f"FROM {table}"
        )

        if self._filters:
            clauses = [
                f"{k} = {format_where_value(v)}" for k, v in self._filters.items()
            ]
            sql += " WHERE " + " AND ".join(clauses)

        sql += f" ORDER BY _score ASC LIMIT {self._top_k};"

        result = orm.query(sql)

        pairs: List[Tuple[Any, float]] = []
        score_idx: Optional[int] = None
        if "_score" in result.columns:
            score_idx = result.columns.index("_score")

        for row in result.rows:
            score = float(row[score_idx]) if score_idx is not None else 0.0
            inst = self._model_cls._from_row(result.columns, row)
            pairs.append((inst, score))

        return pairs
